ArtOfIllusion/src/artofillusion/script/ScriptRunner.java

```diff
@@ -21,9 +21,12 @@
 
 public class ScriptRunner
 {
-  public static final String LANGUAGES[] = {"Groovy", "BeanShell"};
+  public static final String [] LANGUAGES = {"BeanShell", "Groovy"};
+  public static final String [] EXTENSIONS = {"bsh", "groovy"};
```

Comment:
This is... not my favorite way to go about it. The issue is that you now have to remember/track back what index is which, such as lines 48 & 50.

If we've outgrown hard-coded strings here - and I'm not sure that we have - we would want an enum with fields for the name and extension.

Reply:
I'm so happy about this comment! This is really not the way I would have written this myself, but for a lot of things I'm trying to blend into the existing source and coding habits that I see, and I haven't seen a lot of enums anywhere, so I thought I wouldn't introduce new coding practices.. but yes, I 100% agree : enums and/or Maps are obviously the way to go here ; and those static strings should be defined in the *ScriptEngine classes, their values retrieved through methods like getLanguageName() and getFileExtension(), so that new scripting languages implementations (I'm thinking Javascript / Python) would be automatically supported. But I have seen so little interfaces and OO design patterns in the code that I ruled that out.

Reply:
I committed changes so we now use 
> an enum with fields for the name and extension.

..and also a field for the corresponding ScriptEngine class. In most of the code, the language is still implemented as a String (so I have added an UNKNOWN_LANGUAGE constant and used its value wherever I have found a "== null" test), 
and although I would have liked to change things a little more by using the enum value instead of the String wherever possible, I decided to respect the "if it's not broken, don't fix it" coding wisdom and stop here for now. 
As a side note, this change should make implementing a new scripting language easy (I'm thinking javascript with the built-in Nashorn engine).

Reply:
I'm feeling very mixed about the changes in this area:

- On the one hand, I do think that enums are the right solution if we grow to have multiple *built-in* scripting solutions.
- On the other hand, I think that you would need to make a *very* strong case to add other scripting engines to the core application
- If scripting engines are to be supported by plugin, that's a whole different set of API, and enums will *not* work in that case.

How do you see other scripting environments being supported? What are your reasons to want to add them?

Reply:
I agree with all your points, but honestly, I don't really see the issue with adding javascript support (or any other language for that matter) except maybe for code maintenance -even though the code and build process changes involved with javascript look pretty small to me, since it's natively included-. 
Anyway, if you ask me, I would support JS natively with enums, and other languages via plugins, but I'm really not locked in onto any decision right now. 

My reasons to want Javascript (and Python) are long-term projects related to generative art and machine learning.. that, at the current speed things are going (I mean, for me - the time I can spend weekly on AOI has diminished quite a bit since I first started posting on the SF and GH forums :/), are not going to be visible for a long time. So, I'm not going to try hard and push in any direction, and will just go with the flow. 

For now, I would just suggest sticking to enums because it's clean and explicit, and I'll talk again about adding JS (or any other language) again later, when I can give more specific arguments.